### PR TITLE
disable autoClosingBrackets by default everywhere

### PR DESCRIPTION
**Pull request by telamonian**
## References

fixes: #5741

I'll take the approach I outlined in https://github.com/jupyterlab/jupyterlab/issues/5741#issuecomment-744944176

- [x] add `notebook:toggle-autoclosing-brackets` command: toggles autoClosingBrackets option for all cell types
- [x] add a command that combines `fileeditor:toggle-autoclosing-brackets` and `notebook:toggle-autoclosing-brackets`
- [x] update settings menu to use above universal autoClosingBrackets toggle
- [x] disable autoClosingBrackets by default everywhere

## Code changes

Add commands and change menu item as described above

## User-facing changes

autoClosingBrackets will default to false, and the settings menu item will toggle it universally

## Backwards-incompatible changes

NA

### CODE REVIEW

**Comment by jasongrout** [comment]
Bumping to 3.1 since we are releasing the last RC for 3.0 today.

**Comment by ramayer** [comment]
Will this affect auto-closing quotes as well?  They're just as annoying.

**Comment by telamonian** [comment]
@raemayer Yes, the quotes, and all of the other awful auto-pairing inputs we offer. I think I hate it most because it's the only feature we have that will automatically dump garbage text into my work with zero confirmatory input.

Anyhow, you can fix this right now by tweaking some settings in jlab. In the main menus, first click on:

```
settings -> Auto Close Brackets for Text Editor
```

then click on

```
settings -> Advanced Settings Editor
```

then go to the Notebook setting section, and set (all 3 occurences of) `"matchBrackets"` to `false`.

If this seems like a mess, the point of this PR is to condense that sequence down to a single option/menu toggle. I'll definitely make some time for rounding up this PR before 3.1

**Comment by goanpeca** [comment]
@telamonian is this PR still valid?

Thanks!

**Comment by telamonian** [comment]
This has been finished up and rebased, and should now be good to go. Pinging @ellisonbg @jasongrout

**Comment by jasongrout** [comment]
Thanks! Does this also handle the other place editors show up, like the Code Console or the various JSON editors (settings, notebook metadata, etc.)? If not, that's fine - those could be a different PR, as this already moves us quite a way in a positive direction.

It seems like a lot of effort in the PR went into having a single setting that handled editors across different parts of JLab, which makes for a much better user experience, but introduces a lot of complexity in the code. How often do you think a person will want different autoclose behavior in, say, the file editor vs various types of notebook cells? Or different match bracket behavior, etc.? Perhaps we should take a step back and have a single universal editor config used for every editor, and then we can have various parts of jlab override settings as it makes sense, rather than trying to piece together effectively a universal config from many different pieces, like we had to do in this PR because of the current design.

**Comment by telamonian** [comment]
> Code Console or the various JSON editors (settings, notebook metadata, etc.)?

AFAIK there's no way in the present implementation for a user to toggle `autoClosingBrackets` for these "incidental" editors. Currently, `autoClosingBrackets` is always on for incidental editors. With this PR, `autoClosingBrackets` will instead always be off for incidental editors. This is because I changed the default codeeditor config:

https://github.com/jupyterlab/jupyterlab/blob/6b391668e77867b7d3efdf5a317d21948cadd429/packages/codeeditor/src/editor.ts#L796-L814

> How often do you think a person will want different autoclose behavior in, say, the file editor vs various types of notebook cells?

I agree; it doesn't make much sense to segment this option by input widget. Instead, what I might want to see is segmentation of these kinds of options by language, but I don't think we currently have a good way of doing that.

> Perhaps we should take a step back and have a single universal editor config used for every editor

I still think this PR should go in as is for 3.1 since, as you say, it does improve UX. Longer term I agree that universal editor settings make sense to pursue

**Comment by jasongrout** [comment]
I'm testing this, and something seems off, but perhaps it is a problem with master. The codemirror extension has a scrollPastEnd setting that defaults to true. However, in this PR and in master, I don't see the scrollPastEnd behavior in a text editor (where you can scroll so that the last line of the file is at the top of the screen). I do see that the current release (3.0.x) does have scrollPastEnd default behavior.

I was hoping to play a bit with what it would look like with this autoclose setting just being a CodeMirror extension setting, like scrollPastEnd, so it applies to all codemirrors across the application with one setting, rather than having many different settings for each different place codemirror is used. Essentially, a setting that overrides the default.

**Comment by jasongrout** [comment]
Looking at it more closely, it looks like in https://github.com/jupyterlab/jupyterlab/pull/10128/files#diff-ab6be7bde52236b400a3794134a804439f840f5cf2a9bcb325bbe6f4f0b91823R255, the `editorOptions` object is never updated, so the `setOptions` happening in that line never actually does any updating.

**Comment by jasongrout** [comment]
> I was hoping to play a bit with what it would look like with this autoclose setting just being a CodeMirror extension setting, like scrollPastEnd, so it applies to all codemirrors across the application with one setting, rather than having many different settings for each different place codemirror is used. Essentially, a setting that overrides the new default set in this PR.

This is more complicated than I thought originally, in part because we have an awkward dependency relationship between the file editors and the codemirror editors. I created an issue about this at #10352.

So I agree this is probably out of scope for this PR, though fixing this awkward dependency relation would make this PR more elegant (there would be one canonical place for settings and updating editors, instead of multiple places across packages to update when a setting changes, with awkward dependencies like the fileeditor-extension now knowing about notebook commands)

**Comment by jasongrout** [comment]
> Looking at it more closely, it looks like in https://github.com/jupyterlab/jupyterlab/pull/10128/files#diff-ab6be7bde52236b400a3794134a804439f840f5cf2a9bcb325bbe6f4f0b91823R255, the `editorOptions` object is never updated, so the `setOptions` happening in that line never actually does any updating.

Fixed in #10353

File: packages/fileeditor-extension/schema/plugin.json [hunk 1]
```diff
@@ -17,6 +20,12 @@
           "minimum": 1,
           "maximum": 100
         },
+        "handlePaste": {
```
Comment (jasongrout) [inline packages/fileeditor-extension/schema/plugin.json:23]: Did you add this handlePaste option? How does that interact with the code assuming things about how handlePaste is set?

File: packages/fileeditor-extension/schema/plugin.json [hunk 2]
```diff
@@ -62,19 +65,20 @@
       "description": "The configuration for all text editors.\nIf `fontFamily`, `fontSize` or `lineHeight` are `null`,\nvalues from current theme are used.",
       "$ref": "#/definitions/editorConfig",
       "default": {
-        "autoClosingBrackets": true,
+        "autoClosingBrackets": false,
+        "codeFolding": false,
         "fontFamily": null,
         "fontSize": null,
+        "handlePaste": true,
```
Comment (jasongrout) [inline packages/fileeditor-extension/schema/plugin.json:72]: Again, it seems handlePaste was added, but maybe should not be added?

**Review by jasongrout — CHANGES_REQUESTED** [review]
Thanks! Most of this looks good given the system we have in place for editor settings (i.e., given things we can't really change before 4.0). However, adding handlePaste as a setting concerns me. Can you tell why you added it as a setting? It doesn't seem relevant to this PR.

File: packages/fileeditor-extension/schema/plugin.json (hunk 1 above)
Comment (telamonian) [reply packages/fileeditor-extension/schema/plugin.json:23]: In current HEAD, try changing any fileeditor setting in the settings menu: "Increase Text Editor Font Size", etc. You'll see a validation error in the console. I added `handlePaste` to the schema to fix this. If you think this should be fixed differently, let me know

**Comment by krassowski** [comment]
Adding `handlePaste` is needed to fix quite serious bug IMO: https://github.com/jupyterlab/jupyterlab/issues/9411 and https://github.com/jupyterlab/jupyterlab/issues/10489 but maybe it should be moved to another PR so that we can back-port this fix easily.

**Comment by jasongrout** [comment]
> Adding `handlePaste` is needed to fix quite serious bug IMO: #9411 and #10489 but maybe it should be moved to another PR so that we can back-port this fix easily.
> 
> Edit: there might be other ways to fix it, but this one seems fine for me.

The handlePaste setting is not really a setting the user should be setting, I think. I would prefer the approach over in the new PR #10585, which filters the settings for things that are valid, rather than just making any editor config a setting, and removing the handlePaste option from this PR.

File: packages/fileeditor-extension/schema/plugin.json (hunk 1 above)
Comment (jasongrout) [reply packages/fileeditor-extension/schema/plugin.json]: We fixed it in a different way over in #10585.

**Comment by jasongrout** [comment]
It seems that the menu item now only controls the text editor, not the notebook or code console. Investigating to see if there was another merge-with-master problem.

**Comment by jasongrout** [comment]
> It seems that the menu item now only controls the text editor, not the notebook or code console. Investigating to see if there was another merge-with-master problem.

This is now fixed with 8c84729.

I think this is ready for a final review / merge.

The menu item does not adjust the settings for the Code Console (only file editor and notebook), but I think that can be added in a follow-up PR.

**Review by jasongrout — APPROVED** [review]

**Review by blink1073 — APPROVED** [review]
LGTM, thank you both!

**Comment by blink1073** [comment]
Looks like a bunch of the images need updating due to the change in default value

**Comment by jasongrout** [comment]
> Looks like a bunch of the images need updating due to the change in default value

I won't be able to do that in the next 20 minutes before the deadline. :(

**Comment by jasongrout** [comment]
> Looks like a bunch of the images need updating due to the change in default value

huh, several of them seem relevant, but many of them seem like they should not have changed, but do show changes (i.e., that don't have anything to do with the changes in this PR).

**Comment by blink1073** [comment]
I can wait a bit

**Comment by blink1073** [comment]
We should probably finish #10609 and rebase first since 2 of the same images are affected

**Comment by jasongrout** [comment]
> We should probably finish #10609 and rebase first since 2 of the same images are affected

I'd prefer a merge since we already have a merge with conflict resolution, but yeah, if we can just fix the images that are actually affected by the PR, that would be great.

**Comment by blink1073** [comment]
Kicking CI to get a new diff

**Comment by blink1073** [comment]
Failing checks are due to a failed beta 2 release, I am unpublishing those packages now

**Comment by blink1073** [comment]
Unpublished, kicking CI

**Comment by jasongrout** [comment]
@blink1073 - all green!

**Comment by jasongrout** [comment]
#10612 follows up by making similar changes for code consoles. I put it in a follow-up PR since this PR seems to be done at this point.

**Comment by telamonian** [comment]
Thanks @jasongrout for updating and merging this!

**Pull request merged**